Add quotes to font family names

vcanales opened the pull request:
~Checks if the font family name of a given font, either added through Google Fonts or Locally, ends with a number and if so, it adds single quotes to its name.~ Add single quotes to font family names to avoid errors related to trying to parse font names with special characters, white spaces, or numbers.

Also, check if the font family name of a font being uploaded includes non-alphanumeric characters and invalidates the form if so.

Currently, no validation errors are shown in the UI; I'll be making a note of this to add it in as a follow-up change.

Fixes: #468

Commit:
Add quotes to families ending in number on gfonts

Commit:
Add quotes to uploaded font family name

Commit:
Check for special characters on form upload

Commit:
Trim font names

Commit:
Avoid re-rendering due to function definitions

Moves the definition of onFormDataChange inside the effect where it's
used so that it won't be re defined on render.

Also, wraps isFormValid on useCallback for the same reason.

vcanales commented:
I've also added a couple of optimizations that I noticed while fixing this:

1. Refactor promise calls to use async/await syntax, for consistency's sake.
2. Avoid re renders by wrapping `isFormValid` with `useCallback`, and also move the definition of `onFormDataChange` inside the effect where it's used to avoid having to pass it as a dependency.

mikachan commented on src/utils.js:
```diff
@@ -76,3 +76,22 @@ export async function downloadFile( response ) {
+}
+
+/*
+ * Add quotes to fon name if it ends with a number.
```
```suggestion
 * Add quotes to font name if it ends with a number.
```

mikachan commented on src/utils.js:
```diff
@@ -76,3 +76,22 @@ export async function downloadFile( response ) {
+ */
+
+export function addQuotesToName( familyName ) {
+	return trim( familyName.match( /\d$/ ) ? `'${ familyName }'` : familyName );
```
This is only handling font family names that end with a number, so it won't cater to fonts such as [Libre Barcode 128 Text](https://fonts.google.com/specimen/Libre+Barcode+128+Text).

I also noticed that the CSS `font-family` [docs](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#valid_family_names) mention more than just numbers as a requirement for quotes:

> It is a good practice to quote font family names that contain white space, digits, or punctuation characters other than hyphens.

I'm wondering if it would be better to wrap all font names in quotes regardless of whether the name includes numbers or special characters. Are there any downsides to this? 

I had a look through the [LibFont](https://github.com/Pomax/lib-font/) library to see if there were any helpful functions, I didn't find any but I noticed that they're wrapping font names in quotes on output [here](https://github.com/Pomax/lib-font/blob/master/src/utils/fontface.js#L56).

mikachan reviewed (commented):
This works in my testing for the font Exo 2, but I don't think it'll catch the names of the following fonts:

- Libre Barcode 128 Text
- Press Start 2P
- Rock 3D
- Rubik 80s Fade

I've left a longer comment about this below, but I'm wondering if it may be best practice to always wrap font names in quotes, as long as the name isn't a [generic fallback name](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#generic-name) (which they're unlikely to be, as they're custom fonts).

> Refactor promise calls to use async/await syntax
> Avoid re renders by wrapping isFormValid with useCallback,

Nice!

Definitely not a blocker for this PR but it would be great if we could write some tests, especially for things like the utils functions which are hopefully easier to cover in isolation. I know we've not historically done this in this repo - the list of example fonts above made me think that they'd make great test cases.

vcanales commented:
@mikachan You're correct in that those cases should also be considered. Here's a few examples:

```console
// Normal
> new FontFace('Example1', {});
< FontFace { family: "Example1", style: "normal", weight: "normal", stretch: "normal", unicodeRange: "U+0-10FFFF", variant: "normal", featureSettings: "normal", variationSettings: "normal", display: "auto", ascentOverride: "normal" }

// Notice the empty family name:
> new FontFace('Example 2', {});
< FontFace { family: "", style: "normal", weight: "normal", stretch: "normal", unicodeRange: "U+0-10FFFF", variant: "normal", featureSettings: "normal", variationSettings: "normal", display: "auto", ascentOverride: "normal" }

// Same, empty family name:
new FontFace('Rubik 80s Fade', {});
FontFace { family: "", style: "normal", weight: "normal", stretch: "normal", unicodeRange: "U+0-10FFFF", variant: "normal", featureSettings: "normal", variationSettings: "normal", display: "auto", ascentOverride: "normal" }
```

For some reason, fonts that have white spaces and numbers in their name didn't output errors, but they're just as wrong. I'll have `addQuotesToName` just add quotes and trim names without conditions. I can't seem to find any downsides to doing this.

Commit:
Add quotes to all font name strings

https://github.com/WordPress/create-block-theme/pull/476#issuecomment-1812587764

vcanales commented:
> Definitely not a blocker for this PR but it would be great if we could write some tests, especially for things like the utils functions which are hopefully easier to cover in isolation. I know we've not historically done this in this repo - the list of example fonts above made me think that they'd make great test cases.

I don't think we even have the tools for this right now, so I'll add it to my to-do for this week. I agree that this change in particular is very well fit for unit tests.

mikachan commented on src/local-fonts/index.js:
```diff
@@ -33,11 +34,19 @@ function LocalFonts() {
-	};
+	const isFormValid = useCallback( () => {
+		// Check if font name is present and is alphanumeric.
+		const alphanumericRegex = /^[a-z0-9 ]+$/i;
```
Do you think this regex should also include punctuation characters, based on the `font-family` spec? As this would still be a valid font name:

> It is a good practice to quote font family names that contain white space, digits, or punctuation characters other than hyphens.

I'm not sure which characters they specifically are! And I'm also struggling to immediately find many font families that include special characters.

vcanales replied:
I think we just have to make a call, since Font Families can be special characters, unicode, numbers, etc. But for the purpose of generating class names from these font family names, it's practical to request names that are compatible with this requirement. Therefore, restricting these names to alphanumeric characters seems like a sensible approach in this context.

mikachan replied:
> But for the purpose of generating class names from these font family names, it's practical to request names that are compatible with this requirement.

Yup ok, that makes sense 👍

Commit:
Add tests

Add dependencies required for test and linting as well:

- @wordpress/scripts
- @wordpress/eslint-plugin

mikachan reviewed (commented):
We should add the new `addQuotesToName` function to `demoStyle` too (around line 56 of local-fonts/index.js), something like this:

```
		const style = {
			fontFamily: addQuotesToName( formData.name ),
			fontWeight: formData.weight,
			fontStyle: formData.style,
		};
```

This allows the inline font demo on the local fonts page to apply the font family correctly.

Commit:
Temporarily disable eslint rule

Out of scope for this PR; follow-up to come.

vcanales commented:
@mikachan I ended up adding tests here; it was simpler than expected. I added @wordpress/wp-scripts as a dependency now, which undoes #467.

Commit:
Add quotes to demo style font name

mikachan commented:
> I ended up adding tests here; it was simpler than expected.

Awesome! This is a great start. With this we can start adding tests in other areas too!

mikachan reviewed (approved):
Thanks for working on all of this! I think we should bring this in.
